pyroomacoustics/libroom_src/room.cpp

```diff
@@ -276,3 +276,422 @@ int Room::image_source_shoebox(
+			if (min_coord[2] < minZ)
+				minZ = min_coord[2];
+		}	
+	}
+	
+	// If we are in 2D, maxZ = minZ = 0
+	// so there won't be any difference
+	Eigen::Vector3f max_point(maxX, maxY, maxZ);
+	Eigen::Vector3f min_point(minX, minY, minZ);
+	
+	return (min_point - max_point).norm() + 1;
+
+}
+
+
+
+Eigen::VectorXf Room::next_wall_hit(
+					const Eigen::VectorXf &start,
+					const Eigen::VectorXf &end,
+					Eigen::Ref<Vector1i> next_wall_index){
+						
+	/* Computes the next next wall_hit position given a segment defined
+	 * by its endpoints. This method also stores the index of the wall
+	 * that contains this next hit point (we needed to pass it inside a 
+	 * Vector1i to be able to call this function from python).
+	 * 
+	 * If no wall is intersected, then the next_wall_index will be -1*/
+						
+	Eigen::VectorXf result;
+	result.resize(dim);
+	
+	next_wall_index[0] = -1;
+	
+	// Upperbound on the min distance that we could find
+	float min_dist(max_dist);
+	
+	for (size_t i(0); i < walls.size(); ++i){
+		
+		Wall w = walls[i];
+		
+		// To store the result of this iteration
+		Eigen::VectorXf temp_hit;
+		temp_hit.resize(dim);
+
+
+		// As a side effect, temp_hit gets a value (VectorXf) here
+		bool intersects = (w.intersection(start, end, temp_hit) > -1);
+		
+		//~ std::cout << "\n\n\ntemp_hit : " << temp_hit[0]<< " " << temp_hit[1]<< " " << temp_hit[2] << std::endl;
+		//~ std::cout << "start : " << start[0]<< " " << start[1]<< " " << start[2] << std::endl;
+		//~ std::cout << "end : " << end[0]<< " " << end[1]<< " " << end[2] << std::endl;
+
+		if (intersects){
+				
+			
+			
+			float temp_dist = (start-temp_hit).norm();
+			
+			if (temp_dist > libroom_eps and temp_dist < min_dist){
+				
+				min_dist = temp_dist;
+				result = temp_hit;
+				
+				// update the index of the wall that is hit
+				next_wall_index[0] = i;
+			}
+		}
+		
+	}
+	return result;
+}					
+					
+					
+bool Room::scat_ray(const Wall &last_wall,
+					const Eigen::VectorXf &hit_point,
+					float radius,
+					float scat_energy,
+					float travel_time,
+					float time_thres,
+					float sound_speed,
+					std::vector<entry> &output){
+						
+	/*
+	 * Trace a one-hop scattering ray from the last wall hit to the microphone.
+	 * In case the scattering ray can reach the microphone (no wall in
+	 * between), then we add an entry to the output passed by reference.
+	 * 
+    - last_wall: The wall object where last_hit is located
+    - radius : The radius of the circle/sphere microphone
+    - last_hit: An array of length 2 or 3 defining the last wall hit position
+    - scat_energy: The energy of the scattering ray
+    - actual_travel_time: The cumulated travel time of the ray
+    - time_thresh: The time threshold of the ray
+    - sound_speed: The speed of sound
+    * 
+    * Return : true if the scattered ray reached the mic, false otw
+	*/
+	
+	bool result = false;
+	
+	Vector1i next_wall_idx(-1);
+	
+	next_wall_hit(hit_point, mic_pos, next_wall_idx);
```

Comment:
In the case of the scattered ray, the only walls that can intersect with the rays are concave walls. If you think of the case of a convex room, there will never be a wall intersecting the scattered ray, in other word, the microphone is visible from every wall surface. A list of the concave walls indices is maintained in `Room::obstructing walls`. Only these walls need to be checked for intersection.

Reply:
OK nice, I'll take this into account, I thought the obstructing walls variable was only for ISM. I did not want to touch it.

Reply:
When we construct the room, what should the obstructing_walls list contain ? 
Wall objects that are then converted to int through strange casting ? Or just indices of the walls that can be obstructing ? 

Just to know how I should modify this function when we call it for a scattering ray

Reply:
The indices of the walls. They are compute on the phone side

Reply:
Python side!